File: src/content/parseLessonText.ts
import { LessonBlock, PhraseEntry } from "./types";
import {
  extractBracketPronunciation,
  leadingEmoji,
  looksLikeContinuation,
  normalizeAnswer,
  stripLeadingEmoji,
} from "./textUtils";

const STEP_RE = /^Шаг\s+(\d+)\.\s*(.*)$/u;
const DASH_SPLIT = " — ";

// Lesson files may be written either as plain text (lesson 1) or in Markdown
// (lesson 2). These patterns normalise the Markdown flavour into the same
// shapes the plain format already produced, so both render identically.
const HORIZONTAL_RULE = /^\s*([-*_])\1{2,}\s*$/;
const TABLE_SEPARATOR = /^\s*\|?[\s:|-]*-{2,}[\s:|-]*\|?\s*$/;
const TABLE_ROW = /^\s*\|(.+)\|\s*$/;
const HEADING = /^(#{1,6})\s+(.*)$/;
const NUMBERED_HEADING = /^(\d+)\.\s*(.+)$/;

function stripEmphasis(text: string): string {
  return text
    .replace(/\*\*(.+?)\*\*/g, "$1")
    .replace(/__(.+?)__/g, "$1")
    .replace(/(^|\s)\*(\S(?:.*?\S)?)\*(?=\s|$)/g, "$1$2");
}

function tableCells(line: string): string[] {
  const match = line.match(TABLE_ROW);
  if (!match) return [];
  return match[1]
    .split("|")
    .map((cell) => cell.trim())
    .filter((cell) => cell.length > 0);
}

/**
 * Turns the raw lesson text into a sequence of typed blocks for rendering.
 * Wording is never changed — lines are only classified (title, step heading,
 * sub-heading, phrase, or plain line) and Markdown markers are removed.
 */
export function parseLessonText(raw: string): {
  blocks: LessonBlock[];
  phrases: PhraseEntry[];
} {
  const lines = raw
    .split(/\r\n|\n/)
    .map((line) => line.trim())
    .filter((line) => line.length > 0);

  const blocks: LessonBlock[] = [];
  const phrases: PhraseEntry[] = [];
  const seenPhrases = new Set<string>();
  let phraseIndex = 0;
  let titleTaken = false;

  const pushPhrase = (icon: string | undefined, left: string, translation: string) => {
    const { text: german, pronunciation } = extractBracketPronunciation(left);
    blocks.push({ type: "phrase", icon, german, pronunciation, translation });

    // Keyed by normalized text (case/punctuation-insensitive) rather than the
    // raw string, so "Hallo! — Привет!" and a later "Hallo — привет" collapse
    // into one practice-pool entry instead of feeding the exercise generator
    // two options that only differ formally. The displayed card above still
    // shows the raw text exactly as written.
    const key = `${normalizeAnswer(german)}::${normalizeAnswer(translation)}`;
    if (german && translation && !seenPhrases.has(key)) {
      seenPhrases.add(key);
      phrases.push({ id: `phrase-${phraseIndex++}`, german, pronunciation, translation });
    }
  };

  lines.forEach((rawLine, i) => {
    // Structural Markdown noise carries no lesson content.
    if (HORIZONTAL_RULE.test(rawLine) || TABLE_SEPARATOR.test(rawLine)) return;

    const cells = tableCells(rawLine);
    if (cells.length > 0) {
      // The row above a separator is the table's header, not a word pair.
      const nextLine = lines[i + 1];
      const isHeader = nextLine !== undefined && TABLE_SEPARATOR.test(nextLine);
      if (isHeader) {
        blocks.push({ type: "subheading", text: stripEmphasis(cells.join(" · ")) });
        return;
      }
      if (cells.length >= 2) {
        pushPhrase(undefined, stripEmphasis(cells[0]), stripEmphasis(cells.slice(1).join(" ")));
        return;
      }
      blocks.push({ type: "line", text: stripEmphasis(cells[0]) });
      return;
    }

    const headingMatch = rawLine.match(HEADING);
    const line = stripEmphasis(headingMatch ? headingMatch[2].trim() : rawLine);
    if (!line) return;

    if (!titleTaken) {
      titleTaken = true;
      blocks.push({ type: "title", text: line });
      return;
    }

    if (headingMatch) {
      const numbered = line.match(NUMBERED_HEADING);
      if (headingMatch[1].length === 1 && numbered) {
        blocks.push({ type: "step", number: Number(numbered[1]), title: numbered[2] });
        return;
      }
      const icon = leadingEmoji(line);
      blocks.push({ type: "subheading", icon, text: icon ? stripLeadingEmoji(line) : line });
      return;
    }

    const stepMatch = line.match(STEP_RE);
    if (stepMatch) {
      blocks.push({ type: "step", number: Number(stepMatch[1]), title: stepMatch[2] });
      return;
    }

    if (line.includes(DASH_SPLIT)) {
      const dashIndex = line.indexOf(DASH_SPLIT);
      const leftRaw = line.slice(0, dashIndex);
      const translation = line.slice(dashIndex + DASH_SPLIT.length).trim();
      const icon = leadingEmoji(leftRaw);
      pushPhrase(icon, icon ? stripLeadingEmoji(leftRaw) : leftRaw.trim(), translation);
      return;
    }

    const icon = leadingEmoji(line);
    if (icon) {
      blocks.push({ type: "subheading", icon, text: stripLeadingEmoji(line) });
      return;
    }

    blocks.push({ type: "line", text: line, tight: looksLikeContinuation(line) });
  });

  return { blocks, phrases };
}

File: backend/app/legacy_parser/parse_lesson_text.py
"""Exact port of src/content/parseLessonText.ts — see text_utils.py's module
docstring for why bit-for-bit fidelity matters here."""

import regex

from app.legacy_parser.text_utils import (
    clean_quiz_text,
    extract_bracket_pronunciation,
    leading_emoji,
    looks_like_continuation,
    normalize_answer,
    strip_leading_emoji,
)

_STEP_RE = regex.compile(r"^Шаг\s+(\d+)\.\s*(.*)$")
_DASH_SPLIT = " — "

_HORIZONTAL_RULE = regex.compile(r"^\s*([-*_])\1{2,}\s*$")
_TABLE_SEPARATOR = regex.compile(r"^\s*\|?[\s:|-]*-{2,}[\s:|-]*\|?\s*$")
_TABLE_ROW = regex.compile(r"^\s*\|(.+)\|\s*$")
_HEADING = regex.compile(r"^(#{1,6})\s+(.*)$")
_NUMBERED_HEADING = regex.compile(r"^(\d+)\.\s*(.+)$")

_BOLD_STAR = regex.compile(r"\*\*(.+?)\*\*")
_BOLD_UNDERSCORE = regex.compile(r"__(.+?)__")
_ITALIC_STAR = regex.compile(r"(^|\s)\*(\S(?:.*?\S)?)\*(?=\s|$)")


def _strip_emphasis(text: str) -> str:
    text = _BOLD_STAR.sub(r"\1", text)
    text = _BOLD_UNDERSCORE.sub(r"\1", text)
    text = _ITALIC_STAR.sub(r"\1\2", text)
    return text


def _table_cells(line: str) -> list[str]:
    m = _TABLE_ROW.match(line)
    if not m:
        return []
    cells = [c.strip() for c in m.group(1).split("|")]
    return [c for c in cells if c]


def parse_lesson_text(raw: str) -> dict:
    """Turns raw lesson text into a sequence of typed blocks plus a
    deduplicated phrase pool, mirroring parseLessonText.ts's two return
    values exactly (block order and phrase order/ids both matter for the
    diff-verification step)."""
    lines = [line.strip() for line in regex.split(r"\r\n|\n", raw)]
    lines = [line for line in lines if line]

    blocks: list[dict] = []
    phrases: list[dict] = []
    seen_phrases: set[str] = set()
    phrase_index = 0
    title_taken = False

    def push_phrase(icon: str | None, left: str, translation: str) -> None:
        nonlocal phrase_index
        german, pronunciation = extract_bracket_pronunciation(left)
        # The material card keeps the exact text as written.
        blocks.append({"type": "phrase", "icon": icon, "german": german, "pronunciation": pronunciation, "translation": translation})

        key = f"{normalize_answer(german)}::{normalize_answer(translation)}"
        clean_german = clean_quiz_text(german)
        clean_translation = clean_quiz_text(translation)
        if clean_german and clean_translation and key not in seen_phrases:
            seen_phrases.add(key)
            phrases.append(
                {
                    "id": f"phrase-{phrase_index}",
                    "german": clean_german,
                    "pronunciation": pronunciation,
                    "translation": clean_translation,
                }
            )
            phrase_index += 1

    for i, raw_line in enumerate(lines):
        # Structural Markdown noise carries no lesson content.
        if _HORIZONTAL_RULE.match(raw_line) or _TABLE_SEPARATOR.match(raw_line):
            continue

        cells = _table_cells(raw_line)
        if cells:
            next_line = lines[i + 1] if i + 1 < len(lines) else None
            is_header = next_line is not None and _TABLE_SEPARATOR.match(next_line) is not None
            if is_header:
                blocks.append({"type": "subheading", "text": _strip_emphasis(" · ".join(cells))})
                continue
            if len(cells) >= 2:
                push_phrase(None, _strip_emphasis(cells[0]), _strip_emphasis(" ".join(cells[1:])))
                continue
            blocks.append({"type": "line", "text": _strip_emphasis(cells[0])})
            continue

        heading_match = _HEADING.match(raw_line)
        line = _strip_emphasis(heading_match.group(2).strip() if heading_match else raw_line)
        if not line:
            continue

        if not title_taken:
            title_taken = True
            blocks.append({"type": "title", "text": line})
            continue

        if heading_match:
            numbered = _NUMBERED_HEADING.match(line)
            if len(heading_match.group(1)) == 1 and numbered:
                blocks.append({"type": "step", "number": int(numbered.group(1)), "title": numbered.group(2)})
                continue
            icon = leading_emoji(line)
            blocks.append({"type": "subheading", "icon": icon, "text": strip_leading_emoji(line) if icon else line})
            continue

        step_match = _STEP_RE.match(line)
        if step_match:
            blocks.append({"type": "step", "number": int(step_match.group(1)), "title": step_match.group(2)})
            continue

        if _DASH_SPLIT in line:
            dash_index = line.index(_DASH_SPLIT)
            left_raw = line[:dash_index]
            translation = line[dash_index + len(_DASH_SPLIT) :].strip()
            icon = leading_emoji(left_raw)
            push_phrase(icon, strip_leading_emoji(left_raw) if icon else left_raw.strip(), translation)
            continue

        icon = leading_emoji(line)
        if icon:
            blocks.append({"type": "subheading", "icon": icon, "text": strip_leading_emoji(line)})
            continue

        blocks.append({"type": "line", "text": line, "tight": looks_like_continuation(line)})

    return {"blocks": blocks, "phrases": phrases}
